opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -1916,34 +1916,81 @@ namespace Opm {
+                    for (int phase_index = 0; phase_index < np; ++phase_index) {
+                        state.perfPhaseRates()[global_conn_index*np + phase_index] = rst_connection.rates.get(phs[phase_index]);
+                    }
+                    ++local_conn_index;
+                }
+            } else {
+                /*
+                  When we are in this block we are in a situation where some of
+                  the connections have been closed. The connection vectors in
+                  the WellState instance only have storage reserved for the open
+                  connections. Three different sets of connections are
+                  considered in this code:
+
+                    1. The connections from the Schedule::Well class. This set
+                       of connections includes all the connection specified in
+                       the deck; these connection variables cary the state
+                       information.
+
+                    2. The connection results loaded from the restart file. This
+                       connection set has the same size as the connections from
+                       the Schedule object, but these connection object do not
+                       carry state information.
+
+                    3. The connection information in state.perfxxxx(). These
+                       vectors only have storage for open connections.
+
+                  All in all the functionality can be summarized as: use the
+                  Schedule connections to check the active status, for the open
+                  connections the values are assigned from the restart values to
+                  the state.perfxxx() vectors.
+                */
+
+                const auto& sched_connections = getWellEcl(wm.first).getConnections();
+                auto * perf_pressure = state.perfPress().data() + wm.second[1];
+                auto * perf_rates = state.perfRates().data() + wm.second[1];
+                auto * perf_phase_rates = state.perfPhaseRates().data() + wm.second[1]*np;
+                std::size_t local_connection_index = 0;
+                for (std::size_t sched_connection_index = 0; sched_connection_index < sched_connections.size(); sched_connection_index++) {
+                    if (sched_connections[sched_connection_index].state() == Connection::State::OPEN) {
+                        const auto& rst_connection = rst_well.connections[sched_connection_index];
+
+                        perf_pressure[local_connection_index] = rst_connection.pressure;
+                        perf_rates[local_connection_index] = rst_connection.reservoir_rate;
+                        for (int phase_index = 0; phase_index < np; ++phase_index)
+                            perf_phase_rates[local_connection_index*np + phase_index] = rst_connection.rates.get(phs[phase_index]);
+
+                        local_connection_index++;
+                    }
                 }
```

Comment:
If the [`well_perf_data_`](https://github.com/OPM/opm-simulators/blob/061b5c2395e0928885c714c09dbea2447961dc54/opm/simulators/wells/BlackoilWellModel.hpp#L405) is set up by the time we get here, I think you can probably simplify the logic through [`PerforationData::ecl_index`](https://github.com/OPM/opm-simulators/blob/061b5c2395e0928885c714c09dbea2447961dc54/opm/simulators/wells/PerforationData.hpp#L32-L33)

Reply:
Thanks - that was a good suggestion; completely rewritten.